code-dot-org/code-dot-org
# Foorm: Facilitator Rollups Calculation and View

[PLC-816](https://codedotorg.atlassian.net/browse/PLC-816)
Calculate and show per-facilitator rollups for Foorm Surveys. Includes new facilitator rollups tab in summary view as well as showing per-facilitator averages on the workshop rollups page.

New tab view is pictured below (any missing answers are blank). The table contains the following columns:
- Average for this workshop for each facilitator
- Average for each facilitator across all their survey results for this course
- Historic average across all facilitators for this course

The average for each facilitator across all their survey results for this course has also been added to the workshop rollups tab.
![Screen Shot 2020-04-20 at 2 37 09 PM](https://user-images.githubusercontent.com/33666587/79803002-f66b4b00-8315-11ea-942f-0d4545062320.png)

## How to Read this PR

### Rollup Calculation 
- Start with the new file `dashboard/lib/pd/foorm/README.md`. This has a description of the overall path for the survey pipeline calculations, as well as the format the pipeline outputs
- The entry point for the pipeline is `survey_reporter.rb`. This is what calls into the other pipeline functions
- Next look at `rollup_helper.rb`. This takes the rollup questions from `rollups_by_course.json` and adds metadata from the forms to get all the information needed by the rollup creator (title, question choices, type, forms where the question appears, etc.)
- Next look at `rollup_creator.rb`. This does all the calculations to get the rollups. `calculate_averaged_rollup` is the entry point for this file. The rollup is calculated by first getting the sum and count for each question (called the "intermediate rollup"), then getting the average by dividing sum/count.
- Other files on dashboard side are either refactoring or tests

### Rollup View 
Most of the apps files changed contain a change to pass around a facilitators object which maps facilitator ids to names. 
- `results.jsx` sets up the new facilitators tab
- The bulk of real changes on the apps side are in `survey_rollup_table.jsx`. Read the comments at the top of the file and on getTableRows to get an idea of what most of the logic is doing. The two main functions in this file are getTableRows and getTableColumns, which set up the rows/columns for the survey table. The table is made using [reactabular-table](https://reactabular.js.org/#/).

Note: I've added annotations to files via comments on the PR to show changes that are mostly refactoring and don't need to be looked at deeply.





### Future work 
Future steps will be:
- Add frontend tests
- Add role-based views
- More bug bashing
- UI polishing as needed

## Links

- [jira](https://codedotorg.atlassian.net/browse/PLC-816)

## Testing story
Added more unit tests to validate rollups logic on the backend. Tested on the frontend by testing the view at various stages of the survey lifecycle (has no surveys completed, has incomplete data, has complete data). 

### How to Test Locally
The easiest way to test locally end to end is (will need workshop admin permission):
- Make sure Foorms are seeded: `dashboard> bundle exec rake seed:foorm_forms`
- From the rails console, set up a csf workshop with 2 facilitators `workshop = FactoryGirl.create(:csf_workshop, num_facilitators: 2)`, and note the workshop id
- Next add an enrollment to the workshop: `FactoryGirl.create(:pd_enrollment, :from_user, workshop: workshop)` and note the email address given
- Go to http://localhost-studio.code.org:3000/pd/workshop_dashboard/workshops/your-workshop-id and start the workshop, mark the user attended, and end the workshop
- Log in with the teacher's email address and go to http://localhost-studio.code.org:3000/pd/workshop_survey/csf/post101?force_show=true, and fill out the survey (you can go to this url multiple times if force_show=true is set)
- Log back in with your workshop admin account and go to http://localhost-studio.code.org:3000/pd/workshop_dashboard/daily_survey_results_foorm/your-workshop-id to view the survey results

# Reviewer Checklist:
